pkg/sentry/fsimpl/gofer/gofer.go

```diff
@@ -95,7 +95,75 @@ const (
 	cacheRemoteRevalidating  = "remote_revalidating"
 )
 
-const defaultMaxCachedDentries = 1000
+const (
+	defaultMaxCachedDentries         = 1000
+	defaultMaxCachedNegativeChildren = 1000
+)
+
+// +stateify savable
+type negativeNamesCache struct {
+	// namesList stores negative names with fifo list.
+	// name stored in namesList only means it used to be negative
+	// at the moment you pushed it to the list.
+	namesList stringList
+	// negativeCnt count nums of negative names, note that this count may not equal
+	// to length of namesList, because we will not update namesList in DecCount.
+	negativeCnt           uint64
```

Comment:
I think we should keep track of "number of negative children" in `dentry` struct itself. For the following reasons:

1. It nicely complements the existing `dentry.syntheticChildren` counter. You can just add `dentry.negativeChildren`.
2. Having the negative children counter in negative children cache seems counter-intuitive. A cache usually only tracks its cached items and the cache size. The number of negative children is different from cache size. This is the reason we have to add weird methods to the cache API, like `IncCountOrPopBack()` etc. Cache API should be simple: Push(), Pop().
3. Most importantly, this helps us avoid an allocation in the common case where we don't have a cache. As of right now, all directory dentries that call `cacheNewChildLocked()` will end up allocating a new `negativeNamesCache` struct on the heap via `newNegativeNamesCache()`.

Reply:
done